protocol_rfcs/in-commit-timestamps.md

```diff
@@ -0,0 +1,63 @@
+# In-Commit Timestamps
+
+This RFC proposes a new Writer table feature called In-Commit Timestamps. When enabled, commit metadata includes a monotonically increasing timestamp that allows for reliable TIMESTAMP AS OF time travel even if filesystem operations change a commit file's modification timestamp.
+
+**For further discussions about this protocol change, please refer to the Github issue - https://github.com/delta-io/delta/issues/2532**
+
+--------
+
+
+### Commit Provenance Information
+> ***Change to existing section***
+
+A delta file can optionally contain additional provenance information about what higher-level operation was being performed as well as who executed it.
+
+Implementations are free to store any valid JSON [object literal](https://www.w3schools.com/js/js_json_objects.asp) as the `commitInfo` action <ins>unless some table feature (e.g. [In-Commit Timestamps](#in-commit-timestamps)) imposes additional requirements on the data</ins>.
+
+<ins>When In-Commit Timestamp are enabled, writers are required to include a commitInfo action with every commit, which must include the `inCommitTimestamp` field.</ins>
+
+#### Reader Requirements for AddCDCFile
+> ***Change to existing section***
+
+...
+3. Change data readers should return the following extra columns:
+
+Field Name | Data Type | Description
+-|-|-
+_commit_version|`Long`| The table version containing the change. This can be derived from the name of the Delta log file that contains actions.
+_commit_timestamp|`Timestamp`| The timestamp associated when the commit was created. ~~This can be derived from the file modification time of the Delta log file that contains actions.~~ <ins>Depending on whether [In-Commit Timestamps](#in-commit-timestamps) are enabled, this is derived from either the `inCommitTimestamp` field of the `commitInfo` action of the version's Delta log, or from the Delta log's file modification time.</ins>
+
+# In-Commit Timestamps
+> ***New Section after the [Clustered Table](#clustered-table) section***
+
+The In-Commit Timestamps writer feature strongly associates a monotonically increasing timestamp with each commit by storing it in the commit's metadata.
+
+Enablement:
+- The table must be on Writer Version 7.
+- The feature `inCommitTimestamps` must exist in the table `protocol`'s `writerFeatures`.
+- The table property `delta.enableInCommitTimestamps` must be set to `true`.
```

Comment:
What purpose would this table property serve? There isn't much of an enable/disable story, is there? No metadata or data to rewrite, no one-way doors, etc?

CC @bart-samwel since he's thought a _lot_ about properties/upgrade/downgrade issues.

Reply:
>  No metadata or data to rewrite

We might want to remove `delta.inCommitTimestampEnablementTimestamp/Version` from the metadata on disablement. Though it is technically okay even if we do not.

Reply:
I think we would definitely want to remove them, unless we're willing to track multiple ranges of commits where ICT was enabled. But that can easily be done in the same commit that removes the feature from protocol.